File: src/main/java/dev/zprestige/fire/module/movement/anchor/Anchor.java
package dev.zprestige.fire.module.movement.anchor;

import dev.zprestige.fire.event.bus.EventListener;
import dev.zprestige.fire.module.Descriptor;
import dev.zprestige.fire.module.Module;
import dev.zprestige.fire.settings.impl.ComboBox;
import dev.zprestige.fire.settings.impl.Slider;
import dev.zprestige.fire.settings.impl.Switch;
import dev.zprestige.fire.util.impl.Timer;
import net.minecraft.util.math.AxisAlignedBB;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Vec3i;

@Descriptor(description = "Makes getting into holes easier")
public class Anchor extends Module {
    public final ComboBox mode = Menu.ComboBox("Mode", "Stop Motion", new String[]{
            "Stop Motion",
            "Pull",
            "Snap"
    });
    public final Slider snapFall = Menu.Slider("Snap Fall", 1.0f, 0.1f, 5.0f).visibility(z -> mode.GetCombo().equals("Snap"));
    public final Slider pullSpeed = Menu.Slider("Pull Speed", 1.0f, 0.1f, 5.0f).visibility(z -> mode.GetCombo().equals("Pull"));
    public final Slider height = Menu.Slider("Height", 1.0f, 0.1f, 5.0f).visibility(z -> mode.GetCombo().equals("Stop Motion"));
    public final Slider range = Menu.Slider("Range", 1.0f, 0.1f, 5.0f);
    public final Switch onGround = Menu.Switch("On Ground", true).visibility(z -> mode.GetCombo().equals("Snap"));
    protected final Vec3i[] offsets = new Vec3i[]{
            new Vec3i(0, 0, -1),
            new Vec3i(0, 0, 1),
            new Vec3i(-1, 0, 0),
            new Vec3i(1, 0, 0)
    };
    protected final Timer timer = new Timer();
    protected BlockPos pos;

    public Anchor() {
        eventListeners = new EventListener[]{
                new TickListener(this)
        };
    }

    protected void setMovementsFalse() {
        mc.gameSettings.keyBindForward.pressed = false;
        mc.gameSettings.keyBindBack.pressed = false;
        mc.gameSettings.keyBindRight.pressed = false;
        mc.gameSettings.keyBindLeft.pressed = false;
    }

    protected boolean isOver(final BlockPos pos) {
        final AxisAlignedBB bb = mc.player.getEntityBoundingBox();
        for (int i = 0; i < height.GetSlider(); i++) {
            final BlockPos pos1 = pos.up(i);
            final AxisAlignedBB bb1 = new AxisAlignedBB(pos1);
            if (bb.intersects(bb1)) {
                return true;
            }
        }
        return false;
    }

    protected boolean intersects(final BlockPos pos) {
        final AxisAlignedBB bb = mc.player.getEntityBoundingBox();
        for (int i = 0; i < 5; i++) {
            for (final Vec3i vec3i : offsets) {
                final BlockPos pos1 = pos.up(i).add(vec3i);
                final AxisAlignedBB bb1 = new AxisAlignedBB(pos1);
                if (bb.intersects(bb1)) {
                    return false;
                }
            }
        }
        return true;
    }

}

File: src/main/java/dev/zprestige/fire/module/movement/anchor/Anchor.kt
package dev.zprestige.fire.module.movement.anchor

import dev.zprestige.fire.event.bus.EventListener
import dev.zprestige.fire.module.Descriptor
import dev.zprestige.fire.module.Module
import dev.zprestige.fire.settings.impl.ComboBox
import dev.zprestige.fire.settings.impl.Slider
import dev.zprestige.fire.settings.impl.Switch
import dev.zprestige.fire.util.impl.Timer
import net.minecraft.util.math.AxisAlignedBB
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Vec3i

@Descriptor(description = "Makes getting into holes easier")
class Anchor : Module() {
    val mode: ComboBox = Menu.ComboBox("Mode", "Stop Motion", arrayOf(
        "Stop Motion",
        "Pull",
        "Snap"
    )
    )
    val snapFall: Slider = Menu.Slider("Snap Fall", 1.0f, 0.1f, 5.0f).visibility { z: Float? -> mode.GetCombo() == "Snap" }
    val pullSpeed: Slider = Menu.Slider("Pull Speed", 1.0f, 0.1f, 5.0f).visibility { z: Float? -> mode.GetCombo() == "Pull" }
    val height: Slider = Menu.Slider("Height", 1.0f, 0.1f, 5.0f).visibility { z: Float? -> mode.GetCombo() == "Stop Motion" }
    val range: Slider = Menu.Slider("Range", 1.0f, 0.1f, 5.0f)
    val onGround: Switch = Menu.Switch("On Ground", true).visibility { z: Boolean? -> mode.GetCombo() == "Snap" }
    private val offsets = arrayOf(
        Vec3i(0, 0, -1),
        Vec3i(0, 0, 1),
        Vec3i(-1, 0, 0),
        Vec3i(1, 0, 0)
    )
    val timer = Timer()
    var pos: BlockPos? = null

    init {
        eventListeners = arrayOf<EventListener<*, *>>(
            TickListener(this)
        )
    }

    fun setMovementsFalse() {
        mc.gameSettings.keyBindForward.pressed = false
        mc.gameSettings.keyBindBack.pressed = false
        mc.gameSettings.keyBindRight.pressed = false
        mc.gameSettings.keyBindLeft.pressed = false
    }

    fun isOver(pos: BlockPos): Boolean {
        val bb = mc.player.entityBoundingBox
        var i = 0
        while (i < height.GetSlider()) {
            val pos1 = pos.up(i)
            val bb1 = AxisAlignedBB(pos1)
            if (bb.intersects(bb1)) {
                return true
            }
            i++
        }
        return false
    }

    fun intersects(pos: BlockPos): Boolean {
        val bb = mc.player.entityBoundingBox
        for (i in 0..4) {
            for (vec3i in offsets) {
                val pos1 = pos.up(i).add(vec3i)
                val bb1 = AxisAlignedBB(pos1)
                if (bb.intersects(bb1)) {
                    return false
                }
            }
        }
        return true
    }
}